### ISSUE TITLE
Include other languages?

### ISSUE THREAD

**Issue by mpadge** [DavisVaughan/r-tree-sitter#21]
Hi @DavisVaughan, thanks so much for your great tree-sitter work. For an upcoming project at rOpenSci, we aim to use tree-sitter s-expression output to quantify aspects of tree edit distances between codebases, for which this package already has [all the code we need](https://github.com/DavisVaughan/r-tree-sitter/blob/d655234399c9dbf489a57e0416e251ae8b8d99df/src/node.c#L6-L17), but for R only. For our work, we need to  to overcome this limitation:
https://github.com/DavisVaughan/r-tree-sitter/blob/d655234399c9dbf489a57e0416e251ae8b8d99df/README.md?plain=1#L29-L31
and expand the functionality of this package  to the other tree-sitter languages.

Obviously you're thinking about this issue, and working on https://github.com/DavisVaughan/r-tree-sitter/blob/main/R/standalone-language.R. That suggests that there are plans to  expand R bindings to https://github.com/tree-sitter/tree-sitter-c, https://github.com/tree-sitter/tree-sitter-typescript, https://github.com/tree-sitter/tree-sitter-julia, and/or https://github.com/tree-sitter/tree-sitter-python, plus rust via the main repo. Any info on current status and vision would be very much appreciated. Thanks for this extremely useful package, which is already the best-documented and most functionally developed extension outside `api.c`!!

**Comment by DavisVaughan** [comment]
Definitely thinking about and am interested in this.

You're right that `standalone-language.R` is designed so that it can be copied into another grammar's R package bindings as-is. If another grammar exposed a language object using that, it will "just work" with the treesitter package here.

The R package bindings for tree-sitter-r (i.e. the `treesitter.r` package) are nested right in the `bindings/` folder alongside the other bindings, with the idea being that you can do this for other languages too https://github.com/r-lib/tree-sitter-r/tree/main/bindings.

We'd have to get some buy in from the upstream tree-sitter maintainers, since they control the other grammar repos like tree-sitter-c and tree-sitter-python, but I think they'd be willing. We'd probably have to work with them to get some automation utilities going - they are very big on automation since they have so many bindings to do releases for.

The biggest hurdle for them is:
- CRAN releases are still a somewhat manual process, unlike pushing to pypi or crates.io
- CRAN is very picky about what can be in the C code, occasionally requiring patches [like these](https://github.com/search?q=repo%3ADavisVaughan%2Fr-tree-sitter+r-tree-sitter+language%3AC&type=code&l=C). This will come into play for the `scanner.c` files that each grammar can customize

None of this is insurmountable, but will require a decent chunk of work. I'm not sure I have the time to commit to that right now, but if someone else wants to work on adding R bindings for other grammars then you could use the ones for tree-sitter-r as a template to build from (as long as the tree-sitter maintainers want them)

I'd start with tree-sitter-c and tree-sitter-python as those are probably the most useful for R users, which would be released as `treesitter.c` and `treesitter.python` R packages

**Comment by mpadge** [comment]
Thanks for the response @DavisVaughan. I'd  like to suggest I might help develop additional bindings, but have one further issue in that all of the C files, including `api.h`, are in `src/` directories, and so can not be directly linked to in other packages. My personal need is to implement complete tree traversal. This is possible with your implementation here, but only in R, and not in C. I've implemented in C [elsewhere](https://github.com/ropensci-review-tools/pkgsimil/blob/main/src/tree-parser.c), but could only do so by bundling all `tree-sitter` files including the compiled `parser.c` from `tree-sitter generate`. The size of that file alone then completely prevents ability to submit to CRAN. And so the general question is ...

*What can we do to help people directly link to and use `tree-sitter` via `api.h`, and including generated `parser.c` files for R and maybe other languages, within R?*

I'm guessing one plausible solution might be to expand current [`tree-sitter-r/bootstrap.R`](https://github.com/r-lib/tree-sitter-r/blob/main/bindings/r/bootstrap.R) to also copy `parser.c` (tree-sitter-r) and `api.h` (tree-sitter) to local `inst/include` dirs to enable linking? That's also going to (greatly) increase installed size of `tree-sitter-r`, but maybe you could argue the necessity of this with CRAN and get away with it as a one-off that would from that point on allow everybody else to directly link to `api.h` and `parser.c` for R (plus whatever may come in the future). Any thoughts?

**Comment by DavisVaughan** [comment]
> The size of that file alone then completely prevents ability to submit to CRAN

That's not true. The {treesitter.r} package includes the generated `parser.c` file for tree-sitter-r. You can download the source from CRAN and see for yourself https://cran.r-project.org/web/packages/treesitter.r/index.html. It's 5mb, yea, but CRAN is cool with that.

**Comment by DavisVaughan** [comment]
I guess its worth noting that we dropped 250k lines from `parser.c` recently, due to https://github.com/r-lib/tree-sitter-r/pull/132, which dropped it in size significantly. But it was still on CRAN before then too, when `parser.c` was ~13mb

**Comment by DavisVaughan** [comment]
So it sounds like what you are asking for is for:
- This package, {treesitter}, to expose C level bindings to tree-sitter via the [api header](https://github.com/DavisVaughan/r-tree-sitter/blob/main/src/tree-sitter/lib/include/tree_sitter/api.h)
- My other package, {treesitter.r}, to expose the `tree_sitter_r()` function at the very end of the `parser.c` file, i.e. the one with this signature

```
TS_PUBLIC const TSLanguage *tree_sitter_r(void)
```

- So that you could use `LinkingTo` with _both_ of those packages and call all of the tree sitter tooling from C in your own package, bypassing R completely for your needs

**Comment by DavisVaughan** [comment]
What is the scale of what you are trying to do with this? Are you running it on one package? Thousands?

I see https://github.com/ropensci-review-tools/pkgstats/issues/4 and I see that the tree-sitter approach is slower, but maybe there are some improvements we can make to the R side of things before going to this extreme approach of exposing all of the C level details. It would certainly make maintenance harder going forward if we did that right now.

**Comment by DavisVaughan** [comment]
For example, profiling your example from the pkgstats issue on my machine seems to suggest that a few slow places are:
- `check_node()`, which could made made slightly faster 
- `new_node()`, which could be made faster in C
- Using `class = FALSE` for `TreeCursor`, which turns off S3 dispatch in the `$` method, which seemed to help a decent bit actually. This has many downsides as it means you don't have a pretty print method anymore, but we don't use S3 dispatch for anything other than that so maybe we could make it opt in for performance

**Comment by mpadge** [comment]
Thanks @DavisVaughan for the thoughts. My scale comes from needing to apply `pkgstats` to all CRAN packages ever, so currently around 120,000. Current runs are use ctags for tagging, but that's only approximate (and only tags definition statements), whereas tree-sitter is a powerful way to extract exact tags. Potential future applications that we have in mind would extend current CRAN (N~20k) towards O(N^2) realms, so speed would be very critical there.

**Comment by mpadge** [comment]
> So it sounds like what you are asking for is for:
> 
>     * This package, {treesitter}, to expose C level bindings to tree-sitter via the [api header](https://github.com/DavisVaughan/r-tree-sitter/blob/main/src/tree-sitter/lib/include/tree_sitter/api.h?rgh-link-date=2024-09-11T13%3A06%3A22Z)
> 
>     * My other package, {treesitter.r}, to expose the `tree_sitter_r()` function at the very end of the `parser.c` file, i.e. the one with this signature
> 
> 
> ```
> TS_PUBLIC const TSLanguage *tree_sitter_r(void)
> ```
> 
>     * So that you could use `LinkingTo` with _both_ of those packages and call all of the tree sitter tooling from C in your own package, bypassing R completely for your needs

And yes indeed, that would be exactly what I'd like! And if only one of those, then exposing `tree_sitter_r()` from {treesitter.r}. Bundling `api.h` is not too big an issue; it's `grammar.c` that's the problem. Plus exposing `tree_sitter_r()` would then provide a template for future of exposing `tree_sitter_c()`, `tree_sitter_py()`, ...
